### ISSUE TITLE
Google Colab !make fails with "undefined reference to `_glapi_tls_Current'"

### ISSUE THREAD

**Issue by pudepiedj** [googlecolab/colabtools#1215]
Bug report for Colab: http://colab.research.google.com/.

For questions about colab usage, please use [stackoverflow](https://stackoverflow.com/questions/tagged/google-colaboratory/).

- Describe the current behavior:

I am sorry if this is me being stupid but I have looked everywhere I can think of to try to resolve this issue with no success whatever.

In a clean new GPU-runtime session, after loading 

!cuda-install-samples-10.1.sh ~

successfully and

cd /root/NVIDIA_CUDA-10.1_Samples/

I ran

!make.

It succeeds through 0_Simple and 1_Utilities
but part-way through 2_Graphics it fails with
"undefined reference to `_glapi_tls_Current'"
I have been unable to find a fix, and am not sure whether I have failed to do something or this is a bug (it's not easy to know what is pre-installed and what isn't on the Google Colab environment).

If I use a CPU Runtime everything appears to compile, but of course the GPU-dependent compiled code then won't run and gives a "cannot connect to display :0" error on, say, 5_Simulations/particles. Switching to GPU after compilation resets the environment.

- Describe the expected behavior:

Unless I am missing something or not installing something I need, !make should work, either on the whole _Samples directory-set or on single directories requiring GPU and display output. It compiles and the compiled code runs on, say, 0_Simple/matrixMul.
Running !make in a single directory requiring display output, for example 5_Simulations/particles or 5_Simulations/nbody produces the same failure for the same reason.

The only result Googling this error generates is from Gentoo/Linux where there appears to be a sequencing problem over compiling x11-base and xord_server, but I don't know whether this is relevant here and am not sufficiently familiar with the Linux environment to comment.

- The web browser you are using (Chrome, Firefox, Safari, etc.):
Chrome.

- Link to self-contained notebook that reproduces this issue
  (click the *Share* button, then *Get Shareable Link*):
I don't see a *Share* button anywhere but these three lines in a notebook reproduce the error every time if GPU Runtime is enabled:

!cuda-install-samples-10.1.sh ~
cd /root/NVIDIA_CUDA-10.1_Samples/
!make

[image](https://user-images.githubusercontent.com/26062177/81142691-306c5c00-8f68-11ea-868c-4f81258f8b4e.png)

**Comment by colaboratory-team** [comment]
Looks like a bug in NVIDIA_CUDA-10.1_Samples that is worked around by adding a missing lib to the link line. This works for me:
```
!cuda-install-samples-10.1.sh ~ && cd /root/NVIDIA_CUDA-10.1_Samples/ && make GLLINK=-lGLdispatch && echo Yay
```

(the Share button is at the top-right of the colab UI, to the left of the settings gear which is to the left of your avatar / google account switcher)

**Comment by pudepiedj** [comment]
Great! Yes, it compiled everything without any problems.
Thank you!
There are quite a few messages in the !make compilation output indicating
that the NVIDIA_CUDA code is breaking various rules/deprecations etc, so
perhaps it needs a makeover.
Unfortunately, even though it compiles, and all the non-graphical programs
seem to run, it doesn't seem to want to run the graphics programs like
"particles" and "nbody"; gives a message saying "freeglut (): failed to
open display :0". Is it actually possible to run these graphical programs
on Colab? I read somewhere that X-server can't run in a browser, so maybe
it isn't possible. I've tried editing the ssh_config file to allow X11
forwarding, but it doesn't work.

I am so pleased to be able to do (most of) this again and will soon get
going on some other projects that I had to abandon some time ago. My rather
old MacBook has fallen foul of Apple's "tiff" with NVIDIA so the ancient
GForce 750 GPU is no longer supported.
I am loving Google Colab and pleased that (for once) it wasn't me being
stupid that prevented this from working.
